core-bundle/src/Altcha/Altcha.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Altcha;
+
+use Contao\CoreBundle\Altcha\Config\Algorithm;
+use Contao\CoreBundle\Altcha\Exception\InvalidAlgorithmException;
+use Contao\CoreBundle\Entity\Altcha as AltchaEntity;
+use Contao\CoreBundle\Repository\AltchaRepository;
+use Doctrine\ORM\EntityManagerInterface;
+
+/**
+ * @internal
+ */
+class Altcha
+{
+    public function __construct(
+        private readonly AltchaRepository $repository,
+        private readonly EntityManagerInterface $entityManager,
+        private readonly string $secret,
+        private readonly string $altchaAlgorithm,
+        private readonly int $altchaRangeMin,
+        private readonly int $altchaRangeMax,
+        private readonly int $altchaChallengeExpiry,
+    ) {
+    }
+
+    /**
+     * @throws InvalidAlgorithmException
+     */
+    public function createChallenge(string|null $salt = null, int|null $number = null): AltchaChallenge
+    {
+        $algorithms = Algorithm::values();
+
+        if (!\in_array($this->altchaAlgorithm, $algorithms, true)) {
+            throw new InvalidAlgorithmException(sprintf('Invalid algorithm selected. It has to be one of these: %s', implode(', ', $algorithms)));
+        }
+
+        if (!$salt) {
+            $salt = bin2hex(random_bytes(12)).'?expires='.strtotime("now +$this->altchaChallengeExpiry seconds");
+        }
```

Comment:
I was not able to find the place where this `?expires=` parameter gets validated. Shouldn’t there be an `else` block here that parses `?expires=` and checks if the time has passed?

Reply:
The JavaScript handles that:

<img width="306" alt="" src="https://github.com/contao/contao/assets/1192057/6a8a2e24-ed8e-4723-874c-40c7626be993">

To reproduce, change line 49 in the `Altcha` class to `$salt = bin2hex(random_bytes(12)).'?expires='.strtotime("now +5 seconds");`.

Reply:
Yes, but the attacker/spambot will not use our JavaScript ☺️
This would make it possible to do one replay attack every 24 hours.

From the ALTCHA docs:
> The widget will automatically detect the expires parameter. Your server should then verify the expiration during the verification process.

Should be as simple as something like this:

```php
else {
    parse_str(explode('?', $salt, 2)[1] ?? '', $parameters);
    if ((int) ($parameters['expires'] ?? 0) < time()) {
        throw new Exception('…');
    }
}
```